services/beaconchain/src/main/java/tech/pegasys/teku/services/beaconchain/BeaconChainController.java

```diff
@@ -513,12 +520,46 @@ public void onTick(Date date) {
     on_tick(transaction, currentTime);
     transaction.commit().join();
 
-    if (nextSlotDue) {
-      if (syncService.isSyncActive()) {
-        processSlotWhileSyncing();
-      } else {
-        processSlot();
-      }
+    if (nextSlotDue && syncService.isSyncActive()) {
+      processSlotWhileSyncing();
+      return;
+    }
+
+    final UnsignedLong genesisTime = recentChainData.getGenesisTime();
+    final UnsignedLong calculatedSlot =
+        currentTime.compareTo(genesisTime) < 0
+            ? ZERO
+            : ForkChoiceUtil.getCurrentSlot(currentTime, genesisTime);
+    if (currentTime.compareTo(genesisTime) < 0) {
+      return;
+    }
+
+    // tolerate 1 slot difference, not more
+    if (calculatedSlot.compareTo(nodeSlot.getValue().plus(ONE)) > 0) {
+      EVENT_LOG.nodeSlotDriftEvent(nodeSlot.getValue(), calculatedSlot);
+      nodeSlot.setValue(calculatedSlot);
+    }
+
+    final UnsignedLong epoch = compute_epoch_at_slot(nodeSlot.getValue());
+    final UnsignedLong nodeSlotStartTime =
+        ForkChoiceUtil.getSlotStartTime(nodeSlot.getValue(), genesisTime);
+    if (onTickSlotStart == null || calculatedSlot.compareTo(onTickSlotStart) > 0) {
+      onTickSlotStart = nodeSlot.getValue();
+      processSlotStart(epoch);
+    }
+
+    if ((onTickSlotAttestation == null || calculatedSlot.compareTo(onTickSlotAttestation) > 0)
+        && currentTime.compareTo(nodeSlotStartTime.plus(UnsignedLong.valueOf(oneThirdSlotSeconds)))
+            >= 0) {
+      onTickSlotAttestation = nodeSlot.getValue();
+      processSlotAttestation(epoch);
+    }
+    if ((onTickSlotAggregate == null || calculatedSlot.compareTo(onTickSlotAggregate) > 0)
+        && currentTime.compareTo(
+                nodeSlotStartTime.plus(UnsignedLong.valueOf(oneThirdSlotSeconds * 2)))
+            >= 0) {
+      onTickSlotAggregate = nodeSlot.getValue();
+      processSlotAggregate();
     }
```

Comment:
I think I'm following how this all works but it's not easy. Will  take another look in the morning - I suspect that extracting some private methods would help a lot to give names to the various comparisons being done.

Reply:
yep that's why i was considering refactoring, i don't really like it overly, but the next step would be to refactor...